File: gdax/gdax.go
package gdax

import (
	"encoding/json"
	"errors"
	"fmt"
	"sort"
	"strconv"
)

// GetCurrencies list of currencies
func (r *Rest) GetCurrencies() (interface{}, error) {
	body, err := r.publicRequest(get, api+"/currencies")
	if err != nil {
		return nil, err
	}
	var decode interface{}
	err = json.Unmarshal(body, &decode)
	if err != nil {
		return nil, err
	}
	return decode, nil
}

// GetBook map of level 2 product books
func (r *Rest) GetBook(product string) (interface{}, error) {
	body, err := r.publicRequest(get, api+"/products/"+product+"/book?level=2")
	if err != nil {
		return nil, err
	}
	var decode interface{}
	err = json.Unmarshal(body, &decode)
	if err != nil {
		return nil, err
	}
	return decode, nil
}

// GetTicker map of product ticker
func (r *Rest) GetTicker(product string) (interface{}, error) {
	body, err := r.publicRequest(get, api+"/products/"+product+"/ticker")
	if err != nil {
		return nil, err
	}
	var decode interface{}
	err = json.Unmarshal(body, &decode)
	if err != nil {
		return nil, err
	}
	return decode, nil
}

// GetTrades map of product trades
func (r *Rest) GetTrades(product string) (interface{}, error) {
	body, err := r.publicRequest(get, api+"/products/"+product+"/trades")
	if err != nil {
		return nil, err
	}
	var decode interface{}
	err = json.Unmarshal(body, &decode)
	if err != nil {
		return nil, err
	}
	return decode, nil
}

// GetHistory list of candles for product history
func (r *Rest) GetHistory(product, start, end, granularity string) (*CandleList, error) {
	body, err := r.publicRequest(get, api+"/products/"+product+"/candles?start="+start+"&end="+end+"&granularity="+granularity)
	if err != nil {
		return nil, err
	}
	var decode []interface{}
	err = json.Unmarshal(body, &decode)
	if err != nil {
		return nil, err
	}
	candles := make([]*Candle, 0)
	for i := 0; i < len(decode); i++ {
		values, ok := decode[i].([]interface{})
		if !ok {
			return nil, errors.New("not a list")
		}
		candle := &Candle{}
		floatTime, _ := values[0].(float64)
		candle.Time = int64(floatTime)
		candle.Low, _ = values[1].(float64)
		candle.High, _ = values[2].(float64)
		candle.Open, _ = values[3].(float64)
		candle.Closing, _ = values[4].(float64)
		candle.Volume, _ = values[5].(float64)
		candles = append(candles, candle)
	}
	sort.Sort(SortCandlesByTime(candles))
	return &CandleList{product, candles}, nil
}

// GetStats map of 24 hour product statistics
func (r *Rest) GetStats(product string) (interface{}, error) {
	body, err := r.publicRequest(get, api+"/products/"+product+"/stats")
	if err != nil {
		return nil, err
	}
	var decode interface{}
	err = json.Unmarshal(body, &decode)
	if err != nil {
		return nil, err
	}
	return decode, nil
}

// GetAccounts map of account balances
func (r *Rest) GetAccounts() ([]Account, error) {
	body, err := r.privateRequest(get, api, "/accounts", "")
	if err != nil {
		return nil, err
	}
	var decode []interface{}
	err = json.Unmarshal(body, &decode)
	if err != nil {
		var decodeError interface{}
		err = json.Unmarshal(body, &decodeError)
		fmt.Println(decodeError)
		return nil, err
	}
	accounts := make([]Account, 0)
	for i := 0; i < len(decode); i++ {
		values, ok := decode[i].(map[string]interface{})
		if !ok {
			return nil, errors.New("parse error")
		}
		account := Account{}
		account.ID, _ = values["id"].(string)
		account.Currency, _ = values["currency"].(string)
		account.Balance, _ = values["balance"].(float64)
		account.Available, _ = values["available"].(float64)
		account.Hold, _ = values["hold"].(float64)
		account.ProfileID, _ = values["profile_id"].(string)
		accounts = append(accounts, account)
	}
	return accounts, nil
}

// PlaceOrder send a buy or sell order
func (r *Rest) PlaceOrder(rawJs string) (*Order, error) {
	body, err := r.privateRequest(post, api, "/orders", rawJs)
	if err != nil {
		return nil, err
	}
	var decode map[string]interface{}
	err = json.Unmarshal(body, &decode)
	if err != nil {
		return nil, err
	}
	order := &Order{}
	order.ID, _ = decode["id"].(string)
	temp, _ := decode["price"].(string)
	order.Price, _ = strconv.ParseFloat(temp, 64)
	temp, _ = decode["size"].(string)
	order.Size, _ = strconv.ParseFloat(temp, 64)
	order.Product, _ = decode["product_id"].(string)
	order.Side, _ = decode["side"].(string)
	order.Stp, _ = decode["stp"].(string)
	order.Type, _ = decode["type"].(string)
	order.TimeInForce, _ = decode["time_in_force"].(string)
	order.PostOnly, _ = decode["post_only"].(bool)
	order.CreatedAt, _ = decode["created_at"].(string)
	temp, _ = decode["fill_fees"].(string)
	order.FillFees, _ = strconv.ParseFloat(temp, 64)
	temp, _ = decode["fillled_size"].(string)
	order.FilledSize, _ = strconv.ParseFloat(temp, 64)
	temp, _ = decode["executed_value"].(string)
	order.ExecutedValue, _ = strconv.ParseFloat(temp, 64)
	order.Status, _ = decode["status"].(string)
	order.Settled, _ = decode["settled"].(bool)
	return order, nil
}

// ListOrders get open orders
func (r *Rest) ListOrders() (map[string][]*Order, error) {
	body, err := r.privateRequest(get, api, "/orders", "")
	if err != nil {
		return nil, err
	}
	var decode []interface{}
	err = json.Unmarshal(body, &decode)
	if err != nil {
		var decodeError interface{}
		err = json.Unmarshal(body, &decodeError)
		fmt.Println(decodeError)
		return nil, err
	}
	orders := make(map[string][]*Order)
	for i := 0; i < len(decode); i++ {
		values, ok := decode[i].(map[string]interface{})
		if !ok {
			return nil, errors.New("parse error")
		}
		order := &Order{}
		order.ID, _ = values["id"].(string)
		temp, _ := values["price"].(string)
		order.Price, _ = strconv.ParseFloat(temp, 64)
		temp, _ = values["size"].(string)
		order.Size, _ = strconv.ParseFloat(temp, 64)
		order.Product, _ = values["product_id"].(string)
		order.Side, _ = values["side"].(string)
		order.Stp, _ = values["stp"].(string)
		order.Type, _ = values["type"].(string)
		order.TimeInForce, _ = values["time_in_force"].(string)
		order.PostOnly, _ = values["post_only"].(bool)
		order.CreatedAt, _ = values["created_at"].(string)
		temp, _ = values["fill_fees"].(string)
		order.FillFees, _ = strconv.ParseFloat(temp, 64)
		temp, _ = values["fillled_size"].(string)
		order.FilledSize, _ = strconv.ParseFloat(temp, 64)
		temp, _ = values["executed_value"].(string)
		order.ExecutedValue, _ = strconv.ParseFloat(temp, 64)
		order.Status, _ = values["status"].(string)
		order.Settled, _ = values["settled"].(bool)

		if orders[order.Product] == nil {
			orders[order.Product] = make([]*Order, 0)
		}
		orders[order.Product] = append(orders[order.Product], order)
	}
	return orders, nil
}

// GetOrder get an order by id
func (r *Rest) GetOrder(orderID string) (*Order, error) {
	body, err := r.privateRequest(get, api, "/orders/"+orderID, "")
	if err != nil {
		return nil, err
	}
	var decode map[string]interface{}
	err = json.Unmarshal(body, &decode)
	if err != nil {
		return nil, err
	}
	order := &Order{}
	order.ID, _ = decode["id"].(string)
	temp, _ := decode["price"].(string)
	order.Price, _ = strconv.ParseFloat(temp, 64)
	temp, _ = decode["size"].(string)
	order.Size, _ = strconv.ParseFloat(temp, 64)
	order.Product, _ = decode["product_id"].(string)
	order.Side, _ = decode["side"].(string)
	order.Stp, _ = decode["stp"].(string)
	order.Type, _ = decode["type"].(string)
	order.TimeInForce, _ = decode["time_in_force"].(string)
	order.PostOnly, _ = decode["post_only"].(bool)
	order.CreatedAt, _ = decode["created_at"].(string)
	temp, _ = decode["fill_fees"].(string)
	order.FillFees, _ = strconv.ParseFloat(temp, 64)
	temp, _ = decode["fillled_size"].(string)
	order.FilledSize, _ = strconv.ParseFloat(temp, 64)
	temp, _ = decode["executed_value"].(string)
	order.ExecutedValue, _ = strconv.ParseFloat(temp, 64)
	order.Status, _ = decode["status"].(string)
	order.Settled, _ = decode["settled"].(bool)
	return order, nil
}

File: python/gdax.py
from rest import request, private_request
import websocket

SITE = 'api.gdax.com'
FEED_SITE = 'wss://ws-feed.gdax.com'


class Ticker:
    def __init__(self, ticker_data):
        self.trade_id = int(ticker_data.get('trade_id') or 0)
        self.price = float(ticker_data.get('price') or 0)
        self.size = float(ticker_data.get('size') or 0)
        self.bid = float(ticker_data.get('bid') or 0)
        self.ask = float(ticker_data.get('ask') or 0)
        self.volume = float(ticker_data.get('volume') or 0)
        self.time = ticker_data.get('time')


class NewOrder:
    def __init__(self, order_data):
        self.id = order_data.get('id')
        self.price = float(order_data.get('price') or 0)
        self.size = float(order_data.get('size') or 0)
        self.product_id = order_data.get('product_id')
        self.side = order_data.get('side')
        self.stp = order_data.get('stp')
        self.type = order_data.get('type')
        self.time_in_force = order_data.get('time_in_force')
        self.post_only = order_data.get('post_only')
        self.created_at = order_data.get('created_at')
        self.fill_fees = float(order_data.get('fiil_fees') or 0)
        self.filled_size = float(order_data.get('filled_size') or 0)
        self.executed_value = float(order_data.get('executed_value') or 0)
        self.status = order_data.get('status')
        self.settled = order_data.get('settled')


class Order:
    def __init__(self, order_data):
        self.id = order_data.get('id')
        self.size = float(order_data.get('size') or 0)
        self.product_id = order_data.get('product_id')
        self.side = order_data.get('side')
        self.stp = order_data.get('stp')
        self.funds = float(order_data.get('funds') or 0)
        self.specified_funds = float(order_data.get('specified_funds') or 0)
        self.type = order_data.get('type')
        self.post_only = order_data.get('post_only')
        self.created_at = order_data.get('created_at')
        self.done_at = order_data.get('done_at')
        self.done_reason = order_data.get('done_reason')
        self.fill_fees = float(order_data.get('fiil_fees') or 0)
        self.filled_size = float(order_data.get('filled_size') or 0)
        self.executed_value = float(order_data.get('executed_value') or 0)
        self.status = order_data.get('status')
        self.settled = order_data.get('settled')

    def coin_price(self):
        return self.executed_value / self.filled_size

    def profit_price(self):
        margin = (self.specified_funds / self.funds - 1.0) * 2.0 + 1.0
        return self.coin_price() * margin


class Account:
    def __init__(self, account_data):
        self.id = account_data['id']
        self.currency = account_data['currency']
        self.balance = float(account_data['balance'])
        self.available = float(account_data['available'])
        self.hold = float(account_data['hold'])
        self.profile_id = account_data['profile_id']


class Candle:
    def __init__(self, candle_data):
        self.time = int(candle_data[0])
        self.low = float(candle_data[1])
        self.high = float(candle_data[2])
        self.open = float(candle_data[3])
        self.closing = float(candle_data[4])
        self.volume = float(candle_data[5])

    def typical_price(self):
        return (self.high + self.low + self.closing) / 3


def place_order(auth, post):
    read, status = private_request(auth, 'POST', SITE, '/orders', post)
    if status != 200 or not isinstance(read, dict):
        return read, status
    return NewOrder(read), status


def get_order(auth, id):
    read, status = private_request(auth, 'GET', SITE, '/orders/' + id, '')
    if status != 200 or not isinstance(read, dict):
        return read, status
    return Order(read), status


def get_accounts(auth):
    read, status = private_request(auth, 'GET', SITE, '/accounts', '')
    if status != 200 or not isinstance(read, list):
        return read, status
    accounts = {}
    for read_account in read:
        new_account = Account(read_account)
        accounts[new_account.currency] = new_account
    return accounts, status


def get_candles(product, start, end, granularity):
    read, status = request('GET', SITE, '/products/' + product + '/candles?start=' + start + '&end=' + end + '&granularity=' + granularity, '')
    if status != 200 or not isinstance(read, list):
        return read, status
    candles = []
    for read_candle in read:
        candles.append(Candle(read_candle))
    candles.sort(key=lambda c: c.time, reverse=False)
    return candles, status


def get_ticker(product):
    read, status = request('GET', SITE, '/products/' + product + '/ticker', '')
    if status != 200 or not isinstance(read, dict):
        return read, status
    return Ticker(read), status


def get_feed(products, channels):
    feed = websocket.create_connection(FEED_SITE)
    params = {'type': 'subscribe', 'product_ids': products, 'channels': channels}
    feed.send(json.dumps(params))